# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/thebes/gfxPlatform.cpp
# Commit: eac17a862c29
# Full Hash: eac17a862c29a52ae1a389c5e55c0a1cec0db7b6
# Author: Andrew Osmond <aosmond@mozilla.com>
# Date: 2020-01-11 09:48:29
# Regressor Bug: 1600472
# File Overlap Count: 2
# Description:
#   Bug 1600472 - Disable allowing sacrificing subpixel anti-aliasing for small screens. r=jrmuizel
#   
#   This patch only allows sacrificing subpixel anti-aliasing when the
#   screen size is larger than WUXGA, and when the force disable pref is not
#   set. In the future, we may also add disable this for high end GPUs.
# ==============================================================================

diff -r b6e5faa0693d -r eac17a862c29 gfx/thebes/gfxPlatform.cpp
--- a/gfx/thebes/gfxPlatform.cpp	Fri Jan 10 18:24:13 2020 +0000
+++ b/gfx/thebes/gfxPlatform.cpp	Fri Jan 10 16:57:55 2020 +0000
@@ -608,6 +608,10 @@
   gfx::gfxVars::SetWebRenderDebugFlags(flags.bits);
 }
 
+static void WebRenderQualityPrefChangeCallback(const char* aPref, void*) {
+  gfxPlatform::GetPlatform()->UpdateAllowSacrificingSubpixelAA();
+}
+
 #if defined(USE_SKIA)
 static uint32_t GetSkiaGlyphCacheSize() {
   // Only increase font cache size on non-android to save memory.
@@ -2518,6 +2522,15 @@
   }
 }
 
+void gfxPlatform::UpdateAllowSacrificingSubpixelAA() {
+  int64_t kMaxPixels = 1920 * 1200;  // WUXGA
+  bool allowSacrificingSubpixelAA =
+      mScreenPixels > kMaxPixels &&
+      !StaticPrefs::
+          gfx_webrender_quality_force_disable_sacrificing_subpixel_aa();
+  gfxVars::SetAllowSacrificingSubpixelAA(allowSacrificingSubpixelAA);
+}
+
 void gfxPlatform::InitAcceleration() {
   if (sLayersAccelerationPrefsInitialized) {
     return;
@@ -3272,6 +3285,12 @@
     if (XRE_IsParentProcess()) {
       Preferences::RegisterPrefixCallbackAndCall(
           WebRenderDebugPrefChangeCallback, WR_DEBUG_PREF);
+      Preferences::RegisterCallback(
+          WebRenderQualityPrefChangeCallback,
+          nsDependentCString(
+              StaticPrefs::
+                  GetPrefName_gfx_webrender_quality_force_disable_sacrificing_subpixel_aa()));
+      UpdateAllowSacrificingSubpixelAA();
     }
   }
 #if defined(MOZ_WIDGET_GTK)